### PR TITLE
LSIF server metrics

### PR DESCRIPTION
**Pull request by efritz**
**NOTE**: 2100 lines are from a generated grafana dashboard, and a few hundred others are boilerplate for jsonnet.

- Add prometheus instrumentation to the LSIF server
- Create a grafana dashboard for the LSIF server via jsonnet generation
- Add a small POC grafonnet-based utility that can be used to create similar dashboards in the future.

This branch will target master once lsif-clean-sqlite and lsif-linked-reference-results have been merged.

### CODE REVIEW

File: dev/grafana/provisioning/dashboards/sourcegraph/lsif.json [hunk 1]
```diff
@@ -0,0 +1,2086 @@
+{
```
[the remark this answers is not in the document]
Comment (efritz) [reply dev/grafana/provisioning/dashboards/sourcegraph/lsif.json]: Cool. Updated.

File: lsif/src/cache.ts [hunk 2]
```diff
@@ -168,14 +199,21 @@ export class GenericCache<K, V> {
...
+            this.metrics.eventsCounter.labels('hit').inc()
+
             // Ensure entry is locked before returning
+            const entry = node.value
             entry.readers++
             return entry
         }
 
+        // Log cache event
+        this.metrics.eventsCounter.labels('miss').inc()
```
[the remark this answers is not in the document]
Comment (efritz) [reply lsif/src/cache.ts]: No, the node would have been non-falsy in that case (returns entry on line 211).

[the remark this answers is not in the document]
Comment (uwedeportivo) [reply lsif/src/cache.ts]: ah yes, missed the return stmt, sorry :-)

File: lsif/src/server.ts [hunk 3]
```diff
@@ -42,17 +49,29 @@ const STORAGE_ROOT = process.env.LSIF_STORAGE_ROOT || 'lsif-storage'
...
+    app.use(morgan('tiny'))
     app.use(errorHandler)
 
     app.get('/ping', (_, res) => {
         res.send({ pong: 'pong' })
     })
 
+    app.use(
+        promBundle({
+            // TODO - tune histogram buckets or switch to summary
```
Comment (chrismwendt) [inline lsif/src/server.ts]: There's a TODO here, not sure if you intended to resolve it before merging.

Comment (efritz) [reply lsif/src/server.ts]: It's resolved after the worker stuff happens, which requires the metrics here to be split. It was mostly a note to myself to say we need to figure out what acceptable latencies are (but we need to deploy first on order to figure that out).

**Comment by efritz** [comment]
> Can we avoid checking in `lsif.json` and dynamically generate it when it's needed? I'm guessing `lsif.jsonnet` generates `lsif.json`, but I've never done that before. A line comment inside `lsif.jsonnet`, script, or line in a README would be helpful.

There's a readme in that directory that explains how to generate it. I'll pass this idea along to @uwedeportivo who is/will be in charge of most of the (meta/overarching) metrics lifecycle stuff.

**Comment by uwedeportivo** [comment]
for now let's check in the generated json.

File: lsif/src/cache.test.ts [hunk 4]
```diff
@@ -1,7 +1,24 @@
+import promClient from 'prom-client'
 import { GenericCache, createBarrierPromise } from './cache'
 import * as sinon from 'sinon'
 
 describe('GenericCache', () => {
+    const testCacheSizeGauge = new promClient.Gauge({
```
[the remark this answers is not in the document]
Comment (efritz) [reply lsif/src/cache.test.ts]: No, this value just stops undefined values from being deref'd.

File: lsif/src/metrics.ts [hunk 5]
```diff
@@ -0,0 +1,130 @@
+import promClient from 'prom-client'
+
+//
+// Database Metrics
+
+export const DATABASE_QUERY_DURATION_HISTOGRAM = new promClient.Histogram({
```
[the remark this answers is not in the document]
Comment (efritz) [reply lsif/src/metrics.ts]: Will update this in my current branch.

**Pull request merged**